core/chains/evm/logpoller/parser.go

```diff
@@ -0,0 +1,499 @@
+		// set the level to the number 0 and fallthrough to the default case
+		p.ConfirmationLevel = primitives.ConfirmationLevel(0)
+
+		fallthrough
+	default:
+		// the default case passes the confirmation level as a number directly to a subquery
+		v.expression = v.nestedConfQuery(false, uint64(evmtypes.Confirmations(p.ConfirmationLevel)))
+	}
+}
+
+func (v *pgDSLParser) Timestamp(p primitives.Timestamp) {
+	cmp, err := cmpOpToString(p.Operator)
+	if err != nil {
+		v.err = err
+
+		return
+	}
+
+	v.expression = fmt.Sprintf(
+		"%s %s :%s",
+		timestampFieldName,
+		cmp,
+		v.args.withIndexedField(timestampFieldName, time.Unix(int64(p.Timestamp), 0)),
+	)
+}
+
+func (v *pgDSLParser) TxHash(p primitives.TxHash) {
+	bts, err := hexutil.Decode(p.TxHash)
+	if errors.Is(err, hexutil.ErrMissingPrefix) {
+		bts, err = hexutil.Decode("0x" + p.TxHash)
+	}
+
+	if err != nil {
+		v.err = err
+
+		return
+	}
+
+	txHash := common.BytesToHash(bts)
+
+	v.expression = fmt.Sprintf(
+		"%s = :%s",
+		txHashFieldName,
+		v.args.withIndexedField(txHashFieldName, txHash),
+	)
+}
+
+func (v *pgDSLParser) VisitAddressFilter(p *addressFilter) {
+	v.expression = fmt.Sprintf(
+		"address = :%s",
+		v.args.withIndexedField("address", p.address),
+	)
+}
+
+func (v *pgDSLParser) VisitEventSigFilter(p *eventSigFilter) {
+	v.expression = fmt.Sprintf(
+		"%s = :%s",
+		eventSigFieldName,
+		v.args.withIndexedField(eventSigFieldName, p.eventSig),
+	)
+}
+
+func (v *pgDSLParser) nestedConfQuery(finalized bool, confs uint64) string {
+	var (
+		from     = "FROM evm.log_poller_blocks "
+		where    = "WHERE evm_chain_id = :evm_chain_id "
+		order    = "ORDER BY block_number DESC LIMIT 1)"
+		selector string
+	)
+
+	if finalized {
+		selector = "(SELECT finalized_block_number "
+	} else {
+		selector = fmt.Sprintf("(SELECT greatest(block_number - :%s, 0) ",
+			v.args.withIndexedField("confs", confs),
+		)
+	}
+
+	var builder strings.Builder
+
+	builder.WriteString(selector)
+	builder.WriteString(from)
+	builder.WriteString(where)
+	builder.WriteString(order)
+
+	return fmt.Sprintf("%s <= %s", blockFieldName, builder.String())
+}
+
+func (v *pgDSLParser) VisitEventByWordFilter(p *eventByWordFilter) {
+	if len(p.ValueComparers) > 0 {
+		wordIdx := v.args.withIndexedField("word_index", p.WordIndex)
+
+		comps := make([]string, len(p.ValueComparers))
+		for idx, comp := range p.ValueComparers {
+			comps[idx], v.err = makeComp(comp, v.args, "word_value", wordIdx, "substring(data from 32*:%s+1 for 32) %s :%s")
+			if v.err != nil {
+				return
+			}
+		}
+
+		v.expression = strings.Join(comps, " AND ")
+	}
+}
+
+func (v *pgDSLParser) VisitEventTopicsByValueFilter(p *eventByTopicFilter) {
+	if len(p.ValueComparers) > 0 {
+		topicIdx := v.args.withIndexedField("topic_index", p.Topic)
+
+		comps := make([]string, len(p.ValueComparers))
+		for idx, comp := range p.ValueComparers {
+			comps[idx], v.err = makeComp(comp, v.args, "topic_value", topicIdx, "topics[:%s] %s :%s")
```

Comment:
We used to have a validation whether `topicIdx` has values between `[1..3]` or sth like this. I think it should be verified here as well and added to `v.err` field if constraint is violated

Reply:
Postgres starts indexing arrays from 1 afaik, so it should be `topicIdx+1`, right? At least this is how it used to work so far

Reply:
This is structured exactly like the LP function `SelectIndexedLogsTopicGreaterThan`. No modifications are done to the topic index on that one.

Reply:
I see the issue now. Fixed